Vaults merge to master

phpipam opened the pull request:
Hi @GaryAllan ,

since you made crypto class, would you mind reviewing vaults implementation - new vaults branch ?

I used openssl_encrypt/decrypt to store information to DB.

Commit:
Initial commit

phpipam committed:
Merge branch 'vaults' into vaults-1

phpipam committed:
Merge pull request #3158 from phpipam/vaults-1

Vaults update with master code

phpipam commented on app/admin/vaults/unlock-result.php:
```diff
@@ -0,0 +1,45 @@
+}
+else {
+	// write session
+	$_SESSION['vault'.$vault->id] = $_POST['vaultpass'];
```
Hi @GaryAllan , currently key to decrypt is stored to user session in plain text, got better idea ?

GaryAllan replied:
Hi Miha,
First thoughts....If stored in $_SESSION and config.php `$session_storage = "database";` the key is going to be written to the database unprotected. If `$session_storage = "files";` it's going to be written unprotected to a file somewhere, probably under /tmp.

I'll take a look over the next few days. Have you looked at MySQL/MariaDB at rest encryption (database manages the crypto) or client side JavaScript decryption (Browser manages the crypto)?

phpipam replied:
I agree for sessions, there should be some better solution... MariaDB seems complicated for general use with encryption plugin etc.

GaryAllan replied:
How about when a vault is opened we generate a random buffer ($key_a) and store this in $_SESSION and store the decryption key XOR'd with $key_a in a browser cookie set to expire when the browser is closed (cookie requires HTTPS).

If $key_a is random enough, the cookie and $_SESSION data are useless by themselves and can be combined when required for decryption (or use AES encryption instead of XOR, I need to think about it some more).

Anyhow....getting late. Speak tomorrow.

phpipam replied:
Sure, no hurry, thanks !

GaryAllan replied:
Hi Miha,
If there is only one vault it doesn't show the unlock option

![image](https://user-images.githubusercontent.com/18753294/94158409-424c8e00-fe7a-11ea-8e97-3346ff4fd64a.png)

GaryAllan committed:
Fix SCHEMA and upgrade_queries.
Fix get_module_permissions() checks.
Update comments.
i10n fixes.

GaryAllan committed:
Exclude vaultItems.values from logs

Large certificates exceed log.details VARCHAR(1024) capacity
vaultItems.values is encrypted and adds no logging value.

GaryAllan committed:
Fix POST edit-item-certificate-upload-form.php with BASE!=/

GaryAllan committed:
fetch_website_certificate.php

 -Allow fetching of self-signed and internal-CA signed certificates
 -Allow non-standard ports to be specified e.g https://companyapp:8000
 -Fetch the full certificate chain
 -Improve error reporting
 -i10n fixes

vault-item-details-general.php

 -Display full certificate chain

GaryAllan commented:
Hi @phpipam

Can you add `app/admin/vaults/download-certificate.php` it's missing from the PR

Commit:
Added selectable popup masks

Commit:
Added subnet space map

Commit:
Vaults update

Commit:
Merge branch 'vaults' of https://github.com/phpipam/phpipam into vaults

Commit:
Subnet map fixes

Commit:
Updated changelog

Commit:
Updates

Commit:
Updated certificate export

Commit:
Fixed vault unloacking
